src/csharp/Grpc.Tools.CSharpGenerator/PlatformApis.cs

```diff
@@ -0,0 +1,127 @@
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#endregion
+
+// NOTE: This code has been taken from Grpc.Core.Internal so that different
+// codepaths are not used for OS detection. If/when this class becomes
+// public, figure out how to re-use Grpc.Core instead of duplicated code
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace Grpc.Core.Internal
+{
+    /// <summary>
+    /// Utility methods for detecting platform and architecture.
+    /// </summary>
+    internal static class PlatformApis
+    {
+        static readonly bool isLinux;
+        static readonly bool isMacOSX;
+        static readonly bool isWindows;
+        static readonly bool isMono;
+
+        static PlatformApis()
+        {
+#if DNXCORE50
+            isLinux = RuntimeInformation.IsOSPlatform(OSPlatform.Linux);
+            isMacOSX = RuntimeInformation.IsOSPlatform(OSPlatform.OSX);
+            isWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
+#else
+            var platform = Environment.OSVersion.Platform;
+
+            // PlatformID.MacOSX is never returned, commonly used trick is to identify Mac is by using uname.
+            isMacOSX = (platform == PlatformID.Unix && GetUname() == "Darwin");
+            isLinux = (platform == PlatformID.Unix && !isMacOSX);
+            isWindows = (platform == PlatformID.Win32NT || platform == PlatformID.Win32S || platform == PlatformID.Win32Windows);
+#endif
+            isMono = Type.GetType("Mono.Runtime") != null;
+        }
+
+        public static bool IsLinux
+        {
+            get { return isLinux; }
+        }
+
+        public static bool IsMacOSX
+        {
+            get { return isMacOSX; }
+        }
+
+        public static bool IsWindows
+        {
+            get { return isWindows; }
+        }
+
+        public static bool IsMono
+        {
+            get { return isMono; }
+        }
+
+        public static bool Is64Bit
+        {
+            get { return IntPtr.Size == 8; }
+        }
+
+        [DllImport("libc")]
+        static extern int uname(IntPtr buf);
+
+        static string GetUname()

```

Comment:
The idiomatic way to P/Invoke `uname` is use `StringBuilder` as a pointer to mutable string.

```C#
[DllImport("kernel32", CharSet = CharSet.Ansi)]
extern static bool GetComputerNameA(StringBuilder name, ref int size);
	
void Main() {
  int size = 256;
  var name = new StringBuilder { Capacity = size };
  bool ok = GetComputerNameA(name, ref size);
  Console.WriteLine("GetComputerNameA returned {0}, name = {1}, size = {2}", ok, name, size);
}

---
GetComputerNameA returned True, name = KAZOO, size = 5
```
NB `CharSet = CharSet.Ansi` assumes that the output of the function is essentially 7-bit, 1 byte characters, and I am specifically using the narrow-char `GetComputerNameA` version. I think it is safe to assume that about uname (this is what you do with Marshal.PtrToStringAnsi). Anyhow, for the purposes of this code, namely testing whether the name is "Darwin" or nor, this is sufficient, even if you get a fancy UTF-8 encoded name in ancient Babylonian characters. :)